aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -1359,87 +1359,85 @@ void qmaxpool_2d_nhwc_kernel(
-              odata + b * oH * oW * iC + row * oW * iC + col * iC);
-
-          // Loop over reduction block
-          int64_t h_start = row * sH - pH;
-          int64_t w_start = col * sW - pW;
-          int64_t h_end = std::min(h_start + (kH - 1) * dH + 1, iH);
-          int64_t w_end = std::min(w_start + (kW - 1) * dW + 1, iW);
-          while (h_start < 0)
-            h_start += dH;
-          while (w_start < 0)
-            w_start += dW;
-
-          int64_t c = 0;
-
-          // Interleaved vector loop 4x
-          constexpr auto vec_width = Vectorized<scalar_t>::size();
-          for (; c + 4 * vec_width <= iC; c += 4 * vec_width) {
-            Vectorized<scalar_t> acc{
-                scalar_t(std::numeric_limits<scalar_t::underlying>::lowest())};
-            // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
-            Vectorized<scalar_t> accs[4] = {acc, acc, acc, acc};
-            int64_t tcntr = 0;
-            int64_t x, y;
-            for (y = h_start; y < h_end; y += dH) {
-              for (x = w_start; x < w_end; x += dW) {
-                for (const auto i : c10::irange(4)) {
-                  tcntr = y * iW + x;
-                  auto vals = Vectorized<scalar_t>::loadu(
-                      i_p + tcntr * iC + c + Vectorized<scalar_t>::size() * i);
-                  accs[i] = vec::maximum(accs[i], vals);
-                }
-              } // for x
-            } // for y
-            for (const auto i : c10::irange(4)) {
-              accs[i].store(o_p + c + Vectorized<scalar_t>::size() * i);
-            }
-          } // for c
-
-          // Vector loop
-          for (; c + vec_width <= iC; c += vec_width) {
-            Vectorized<scalar_t> acc{
-                scalar_t(std::numeric_limits<scalar_t::underlying>::lowest())};
-            int64_t tcntr = 0;
-            int64_t x, y;
-            for (y = h_start; y < h_end; y += dH) {
-              for (x = w_start; x < w_end; x += dW) {
-                tcntr = y * iW + x;
-                auto vals = Vectorized<scalar_t>::loadu(i_p + tcntr * iC + c);
-                acc = vec::maximum(acc, vals);
-              } // for x
-            } // for y
-            acc.store(o_p + c);
-          } // for c
-
-          for (; c < iC; ++c) {
-            auto max_val = std::numeric_limits<scalar_t::underlying>::lowest();
-            int64_t tcntr = 0;
-            int64_t x, y;
-            for (y = h_start; y < h_end; y += dH) {
-              for (x = w_start; x < w_end; x += dW) {
+    int64_t nBatch = qx.size(0);
+    at::parallel_for(0, nBatch * oH * oW, 0, [&](int64_t begin, int64_t end) {
+      int64_t b{0}, row{0}, col{0};
+      data_index_init(begin, b, nBatch, row, oH, col, oW);
+
+      for (const auto i : c10::irange(begin, end)) {
+        auto* i_p = reinterpret_cast<scalar_t::underlying*>(idata + b * iW * iH * iC);
+        auto* o_p = reinterpret_cast<scalar_t::underlying*>(odata + i * iC);
```

Comment:
@dzdang, as I understand this kernel doesn't allocate space when iterating each input HW plane when updating corresponding value on the output index.

the above are calculating input and start output index of each (n, h, w) and start input index of (n).

usually on `channels last` memory format we can have two types of implementation for pooling kernels, since we are vectoring on `channels`:
1. iterate each input lane of channels at a time and store the immediate values on output lane of channels, advantage is memory access is contiguous, disadvantage is need to extra writes to output. when scalar_t != acc_scalar_t, need to allocate extra buffer to keep the immediate values on acc_scalar_t. for example the current [fp32](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/AvgPoolKernel.cpp#L99) and [bf16](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/AvgPoolKernel.cpp#L247) kernels use this approach, the temp buffer is thread local and reused across output (n, h, w).
2. iterate each input lane of channels at a block at a time, this kernel uses this approach, the block is 4 * vec_length. advantage is that no extra write to output since it is using __m256i to store the immediate values, disadvantage is the memory access pattern is non-contiguous (when moving to next block, the stride is C).

Reply:
by the way, this PR is only change the kernel from sequential to parallel, noting else is changed.